[fx/package] make GraphModules packageable

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51489 [fx/package] make GraphModules packageable**
* #51420 [fx] Make imports more selective in imported code.
* #51826 [package] make torch.package produce unified format
* #51595 [package] Use custom persistent_load in PackageImporter
* #51694 [package] more reliable method for determining standard library-ness

FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

suo committed:
[fx/package] make GraphModules packageable

Okay, this one is really sad. Basically:

FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
- `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
- Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

This PR introduces some explicit cooperation between `torch.package` and
`torch.fx` to work around these problems.

To `torch.package`, we add the concept of the "current"
importer/exporter, a piece of global state that's checked when
loading/saving `GraphModule`s.

For `GraphModule` , we make each `GraphModule` instance aware of the
`PackageImporter` that created it (if there is one), so that it can
correctly recompile the Python code with imports that reference a
package.

Alternatives that I considered:
- Figure out a better way to serialize GraphModules in packages.
`to_folder` seems promising, but I can't figure out how to get the
loading path to "just work" without doing some surgery on the Pickle
bytecode, which is probably worse than this approach.
- Use `persistent_id` and `persistent_load` to store the Python code.
This doesn't work because GraphModules have code + data, and
`persistent_id` would force us to take over serialization of the data as
well.
- Don't serialize GraphModules through code, and have a special format
that can handle
  1. references to external Python types outside the graph (e.g. `import
  foo`s), and
  2. the FX IR structure.

  This has the disadvantage of introducing a whole new format and
  serialization scheme that is bound to be less fleixble than saving
  code. It also doesn't fundamentally solve the problem: we still need
  to know where to look to resolve references to external types, which
  requires that the deserializtion part know about the routine.

I am very open to better ways to do this, but I couldn't really think of
any :(

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


Okay, this one is really sad. Basically:

FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
- `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
- Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

This PR introduces some explicit cooperation between `torch.package` and
`torch.fx` to work around these problems.

To `torch.package`, we add the concept of the "current"
importer/exporter, a piece of global state that's checked when
loading/saving `GraphModule`s.

For `GraphModule` , we make each `GraphModule` instance aware of the
`PackageImporter` that created it (if there is one), so that it can
correctly recompile the Python code with imports that reference a
package.

Alternatives that I considered:
- Figure out a better way to serialize GraphModules in packages.
`to_folder` seems promising, but I can't figure out how to get the
loading path to "just work" without doing some surgery on the Pickle
bytecode, which is probably worse than this approach.
- Use `persistent_id` and `persistent_load` to store the Python code.
This doesn't work because GraphModules have code + data, and
`persistent_id` would force us to take over serialization of the data as
well.
- Don't serialize GraphModules through code, and have a special format
that can handle
  1. references to external Python types outside the graph (e.g. `import
  foo`s), and
  2. the FX IR structure.

  This has the disadvantage of introducing a whole new format and
  serialization scheme that is bound to be less fleixble than saving
  code. It also doesn't fundamentally solve the problem: we still need
  to know where to look to resolve references to external types, which
  requires that the deserializtion part know about the routine.

I am very open to better ways to do this, but I couldn't really think of
any :(

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


Okay, this one is really sad. Basically:

FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
- `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
- Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

This PR introduces some explicit cooperation between `torch.package` and
`torch.fx` to work around these problems.

To `torch.package`, we add the concept of the "current"
importer/exporter, a piece of global state that's checked when
loading/saving `GraphModule`s.

For `GraphModule` , we make each `GraphModule` instance aware of the
`PackageImporter` that created it (if there is one), so that it can
correctly recompile the Python code with imports that reference a
package.

Alternatives that I considered:
- Figure out a better way to serialize GraphModules in packages.
`to_folder` seems promising, but I can't figure out how to get the
loading path to "just work" without doing some surgery on the Pickle
bytecode, which is probably worse than this approach.
- Use `persistent_id` and `persistent_load` to store the Python code.
This doesn't work because GraphModules have code + data, and
`persistent_id` would force us to take over serialization of the data as
well.
- Don't serialize GraphModules through code, and have a special format
that can handle
  1. references to external Python types outside the graph (e.g. `import
  foo`s), and
  2. the FX IR structure.

  This has the disadvantage of introducing a whole new format and
  serialization scheme that is bound to be less fleixble than saving
  code. It also doesn't fundamentally solve the problem: we still need
  to know where to look to resolve references to external types, which
  requires that the deserializtion part know about the routine.

I am very open to better ways to do this, but I couldn't really think of
any :(

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

jamesr66a commented:
FYI I want to wait until after the release branch cut to review/merge this (Friday)

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

This PR introduces some new concepts to torch.package to fix these
problem.

- `ImporterProxy`: When unpickled by a PackageImporter, this object will
trun into the PackageImporter instance itself. This lets `__reduce__`
function authors access the current PackageImporter instance in its
reduction function. This is accomplished by `persistent_id` hackery.
This solves problem #1 above.
- `__torch_package_persist__`: `persistent_id` in PackageExporter will
check for this method and run it if it exists, passing in the
PacakgeExporter itself as an argument. We use this basically as a hook
function that runs whenenver a GraphModule instance is pickled.  This
hook registers the necessary imports, solving problem #2 above.

There are some follow ups, in particular, you can't really return a
non-None value from `__torch_package_persist__` because we don't
currently have a way to hook into `persistent_load` on the other side,
so you're limited to running side-effectful hooks.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

This PR introduces some new concepts to torch.package to fix these
problem.

- `ImporterProxy`: When unpickled by a PackageImporter, this object will
trun into the PackageImporter instance itself. This lets `__reduce__`
function authors access the current PackageImporter instance in its
reduction function. This is accomplished by `persistent_id` hackery.
This solves problem #1 above.
- `__torch_package_persist__`: `persistent_id` in PackageExporter will
check for this method and run it if it exists, passing in the
PacakgeExporter itself as an argument. We use this basically as a hook
function that runs whenenver a GraphModule instance is pickled.  This
hook registers the necessary imports, solving problem #2 above.

There are some follow ups, in particular, you can't really return a
non-None value from `__torch_package_persist__` because we don't
currently have a way to hook into `persistent_load` on the other side,
so you're limited to running side-effectful hooks.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

This PR introduces some new concepts to torch.package to fix these
problem.

- `ImporterProxy`: When unpickled by a PackageImporter, this object will
trun into the PackageImporter instance itself. This lets `__reduce__`
function authors access the current PackageImporter instance in its
reduction function. This is accomplished by `persistent_id` hackery.
This solves problem #1 above.
- `__torch_package_persist__`: `persistent_id` in PackageExporter will
check for this method and run it if it exists, passing in the
PacakgeExporter itself as an argument. We use this basically as a hook
function that runs whenenver a GraphModule instance is pickled.  This
hook registers the necessary imports, solving problem #2 above.

There are some follow ups, in particular, you can't really return a
non-None value from `__torch_package_persist__` because we don't
currently have a way to hook into `persistent_load` on the other side,
so you're limited to running side-effectful hooks.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo commented on torch/fx/graph_module.py:
```diff
@@ -320,7 +355,7 @@ def __reduce__(self):
         dict_without_graph = self.__dict__.copy()
         del dict_without_graph['_graph']
-        return (deserialize_graphmodule, (dict_without_graph,))
+        return (deserialize_graphmodule, (ImporterProxy(), dict_without_graph,))
```
If we do this, we cannot use the C pickler, right?

zdevito replied:
I think we can still use the C pickler: use persistent_id to intercept pickling, check for `__reduce_package` key, and if present ,return `('package_pickle', _user_reduce_value_)` as the persistent key. Undo in persistent load on the other side.

suo commented on torch/fx/graph_module.py:
```diff
@@ -310,6 +325,26 @@ def wrapped_call(self, *args, **kwargs):
+        # GraphModule pickling.
+        #
+        # GraphModule needs to register its dependencies in PackageExporter, so
+        # do it here.
```
hmmmmmmmmmmmmm I think I tried this and it doesn't work, but I can't remember why. So I'll try it again and see :P

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

This PR introduces some new concepts to torch.package to fix these
problem.

- `ImporterProxy`: When unpickled by a PackageImporter, this object will
trun into the PackageImporter instance itself. This lets `__reduce__`
function authors access the current PackageImporter instance in its
reduction function. This is accomplished by `persistent_id` hackery.
This solves problem #1 above.
- `__torch_package_persist__`: `persistent_id` in PackageExporter will
check for this method and run it if it exists, passing in the
PacakgeExporter itself as an argument. We use this basically as a hook
function that runs whenenver a GraphModule instance is pickled.  This
hook registers the necessary imports, solving problem #2 above.

There are some follow ups, in particular, you can't really return a
non-None value from `__torch_package_persist__` because we don't
currently have a way to hook into `persistent_load` on the other side,
so you're limited to running side-effectful hooks.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

This PR introduces some new concepts to torch.package to fix these
problem.

- `ImporterProxy`: When unpickled by a PackageImporter, this object will
trun into the PackageImporter instance itself. This lets `__reduce__`
function authors access the current PackageImporter instance in its
reduction function. This is accomplished by `persistent_id` hackery.
This solves problem #1 above.
- `__torch_package_persist__`: `persistent_id` in PackageExporter will
check for this method and run it if it exists, passing in the
PacakgeExporter itself as an argument. We use this basically as a hook
function that runs whenenver a GraphModule instance is pickled.  This
hook registers the necessary imports, solving problem #2 above.

There are some follow ups, in particular, you can't really return a
non-None value from `__torch_package_persist__` because we don't
currently have a way to hook into `persistent_load` on the other side,
so you're limited to running side-effectful hooks.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

zdevito commented on torch/fx/graph.py:
```diff
@@ -707,15 +722,37 @@ def emit_node(node : Node):
+        # This is to detect collisions.
+        demangled_names: Dict[str, str] = {}
+
+        def demangle(name):
```
Managing leaking into fx indicates something is not right here. Mangling shouldn't leak at all. Consider using the same approach for any other package code being re-packaged: the exporter needs to know about the importers symbols to successfully package it.

zdevito commented on torch/fx/graph_module.py:
```diff
@@ -56,12 +70,20 @@ def __init__(self, body):
+        # Otherwise, we're unpickling from a regular pickle file. Retrieve the
+        # serialized forward that is stashed in the object state.
+        try:
+            CodeOnlyModule.forward = _forward_from_src(body['_code'], importer)
```
I know it was here before, but copypasta with unnecessary exception: `body.get('_code') or body['code']`

zdevito commented on torch/fx/graph_module.py:
```diff
@@ -35,14 +36,27 @@ def patched_getline(*args, **kwargs):
+    return _deserialize_graph_module(importer, body)
+
+
+def _deserialize_graph_module(importer: Optional[PackageImporter], body: dict) -> torch.nn.Module:
```
This funnels two paths through the same function where the very first line is an if-statement that splits the paths again. Makes more sense to put the code that loads forward into these functions and have `_deserialize_..` take it as an argument.

zdevito commented on torch/fx/graph_module.py:
```diff
@@ -310,6 +337,15 @@ def wrapped_call(self, *args, **kwargs):
+    def __reduce_package__(self, exporter: PackageExporter):
+        module_id = exporter.get_unique_id()
+        exporter.save_source_string(f'fx-generated._{module_id}', self.code)
+        self._fx_generated_id = module_id
```
This never gets deleted on the other end. Might make more sense to keep this as a separate argument. Only reason it is getting put in self here is because of trying to match the arguments to the other deserialize call.

zdevito commented on torch/package/importer.py:
```diff
@@ -160,13 +160,19 @@ def persistent_load(saved_id):
+                    load_tensor(data_type, size, key, _maybe_decode_ascii(location), restore_location)
+                storage = loaded_storages[key]
+                return storage
+            elif typename == 'package-importer':
```
We don't need this anymore right?

suo replied:
oop

zdevito reviewed (commented):
I have a few comments, but most of the things look good. We need to resolve how mangling is happening here, because it is a good sign that our current approach will hit issues.

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo committed:
Update on "[fx/package] make GraphModules packageable"


FX serializes things by serializing Python code as a string and exec'ing
it on load. This accomplishes one goal (we don't have to pickle the
graph object directly) but breaks the pickle abstraction in ways that
are not composable with `torch.package`.

In particular:
1. `forward` is serialized by saving Python code. On load, it's installed
by  `exec`ing that code. This `exec` call needs to have the right
importer installed, otherwise it will not import modules from the
`torch.package` but instead import from the Python environment.
2. Any types/functions used are emitted as `import` statement in the
generated Python code. These are effectively dynamic dependencies of the
`GraphModule` being saved, and need to be registered as such so that the
`PackageImporter` will package them.

To address these, this PR introduces a new protocol for the
importer/exporter: `__reduce_package__`.

A class can implement `__reduce_package__` to customize how it is placed
in the importer/exproter. It functions very similarly to `__reduce__`,
except:
- `__reduce_package__` takes one argument, which is the `PackageExporter`
instance. Users can use this instance to save stuff to the package to
implement their serialization. `__reduce__` takes no args.
- Only the 2-element tuple version of the return value for `__reduce__`
is supported (this could be extended if necessary).
- When the reduction function is called on load, an additional argument
is added to the beginning of the args tuple. This is the `PackageImporter`
instance doing the loading.

The `__reduce_package__` protocol is defined using `persistent_id` and
`persistent_load`, which ensures that we can still use the cpickle
implementation of the pickler by default.

Differential Revision: [D26182082](https://our.internmc.facebook.com/intern/diff/D26182082)

[ghstack-poisoned]

suo commented:
closing in favor of https://github.com/pytorch/pytorch/pull/51971

jamesr66a commented:
@suo  is your last comment backwards?
